pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageBrokerRequestHandler.java

```diff
@@ -0,0 +1,204 @@
+  private final String _reducerHostname;
+  private final int _reducerPort;
+
+  private final MailboxService<Mailbox.MailboxContent> _mailboxService;
+  private QueryEnvironment _queryEnvironment;
+  private QueryDispatcher _queryDispatcher;
+
+  public MultiStageBrokerRequestHandler(PinotConfiguration config, BrokerRoutingManager routingManager,
+      AccessControlFactory accessControlFactory, QueryQuotaManager queryQuotaManager, TableCache tableCache,
+      BrokerMetrics brokerMetrics) {
+    super(config, routingManager, accessControlFactory, queryQuotaManager, tableCache, brokerMetrics);
+    LOGGER.info("Starting Multi-stage BrokerRequestHandler.");
+    String reducerHostname = config.getProperty(QueryConfig.KEY_OF_QUERY_RUNNER_HOSTNAME);
+    if (reducerHostname == null) {
+      // use broker ID as host name, but remove the
+      String brokerId = config.getProperty(CommonConstants.Broker.CONFIG_OF_BROKER_ID);
+      brokerId = brokerId.startsWith(CommonConstants.Helix.PREFIX_OF_BROKER_INSTANCE) ? brokerId.substring(
+          CommonConstants.Helix.SERVER_INSTANCE_PREFIX_LENGTH) : brokerId;
+      brokerId = StringUtils.split(brokerId, "_").length > 1 ? StringUtils.split(brokerId, "_")[0] : brokerId;
+      reducerHostname = brokerId;
+    }
+    _reducerHostname = reducerHostname;
+    _reducerPort = config.getProperty(QueryConfig.KEY_OF_QUERY_RUNNER_PORT, QueryConfig.DEFAULT_QUERY_RUNNER_PORT);
+    _queryEnvironment = new QueryEnvironment(new TypeFactory(new TypeSystem()),
+        CalciteSchemaBuilder.asRootSchema(new PinotCatalog(tableCache)),
+        new WorkerManager(_reducerHostname, _reducerPort, routingManager));
+    _queryDispatcher = new QueryDispatcher();
+    _mailboxService = new GrpcMailboxService(_reducerHostname, _reducerPort);
+
+    // TODO: move this to a startUp() function.
+    _mailboxService.start();
+  }
+
+  @Override
+  public BrokerResponseNative handleRequest(JsonNode request, @Nullable RequesterIdentity requesterIdentity,
+      RequestContext requestContext)
+      throws Exception {
+    long requestId = _requestIdGenerator.incrementAndGet();
+    requestContext.setBrokerId(_brokerId);
+    requestContext.setRequestId(requestId);
+    requestContext.setRequestArrivalTimeMillis(System.currentTimeMillis());
+
+    // First-stage access control to prevent unauthenticated requests from using up resources. Secondary table-level
+    // check comes later.
+    boolean hasAccess = _accessControlFactory.create().hasAccess(requesterIdentity);
+    if (!hasAccess) {
+      _brokerMetrics.addMeteredGlobalValue(BrokerMeter.REQUEST_DROPPED_DUE_TO_ACCESS_ERROR, 1);
+      LOGGER.info("Access denied for requestId {}", requestId);
+      requestContext.setErrorCode(QueryException.ACCESS_DENIED_ERROR_CODE);
+      return new BrokerResponseNative(QueryException.ACCESS_DENIED_ERROR);
+    }
+
+    if (request.has(CommonConstants.Broker.Request.SQL)) {
```

Comment:
To avoid a call to both `request.has` and request.get(), we can do 

```
JsonNode sql = request.get(Broker.Request.SQL);
if (sql == null) {
   throw new BadQueryRequestException("Failed to find 'sql' in the request: " + request);
} else {
   return handleSQLRequest(.....)
}
```